src/include/sof/audio/buffer.h

```diff
@@ -164,6 +164,32 @@ void comp_update_buffer_produce(struct comp_buffer *buffer, uint32_t bytes);
 /* called by a component after consuming data from this buffer */
 void comp_update_buffer_consume(struct comp_buffer *buffer, uint32_t bytes);
 
+/**
+ * Invalidates control block of a buffer to ensure current state and params
+ * readout.
+ * @param buffer Buffer component to invalidate
+ */
+static inline void buffer_control_invalidate(struct comp_buffer *buffer)
```

Comment:
See @akloniex  - he is fixing the same issue and it's more advanced with simpler public API.

Reply:
As far as I know @akloniex is rather changing api for process component (especially usage in copy function), for something with higher level of abstraction. This PR is focused on component fix, and because of specific of smart_amp I guess that here, this low-level routines will be needed anymore - I use this functions in prepare() function.

Reply:
Yes that's rigght, but this looks like we are also touchng buffer locking and buffer invalidation/writeback which @akloniex  is also doing and rolling into his PR.

Reply:
Buffer control invalidation and writeback are also extracted by @akloniex in his PR. If you don't want to wait for it, you should use buffer_lock/unlock functions as it's done in other components.

Reply:
Updated with lock/unlock solution

Reply:
@lgirdwood Since the fate of PR https://github.com/thesofproject/sof/pull/3042 is uncertain, as we discussed on last meeting, I would recommend going with current approach with buffer locking/unlocking, that is consistent with current implementation on master.